tools/golang/pkg/encryption/encryption_test.go

```diff
@@ -0,0 +1,134 @@
+		"outputPrefixType":"CRUNCHY",
+		"status":"ENABLED"
+	}
+	],
+	"primaryKeyId":3962548922
+}`
+
+func loadTestFileString(filename string) string {
+	path := filepath.Join("testdata", filename) // relative path
+	bytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return ""
+	}
+	return string(bytes)
+}
+
+func TestGetTinkPublicKeySuccess(t *testing.T) {
+	httpServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			t.Errorf("Download request was method %s; want GET", r.Method)
+		}
+
+		n, err := w.Write([]byte(googPublicKeyStr))
+		if err != nil {
+			t.Errorf("Failed to write fake response: %v", err)
+		}
+		if n != len([]byte(googPublicKeyStr)) {
+			t.Errorf("Wrote %d bytes of fake response; want %d", n, len([]byte(googPublicKeyStr)))
+		}
+	}))
+	defer httpServer.Close()
+	pubKey, err := RetrieveTinkPublicKey(httpServer.URL)
+	if err != nil {
+		t.Errorf("Failed to retrieve Tink public key.")
+	}
+	if pubKey.PrimaryKeyId != 3962548922 {
+		t.Errorf("Invalid primary key ID: %d", pubKey.PrimaryKeyId)
+	}
+	if pubKey.Key[0].KeyId != 3962548922 {
+		t.Errorf("Invalid key ID found: %d", pubKey.Key[0].KeyId)
+	}
+}
+
+func TestGetTinkPublicKeyGetFailure(t *testing.T) {
+	httpServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			t.Errorf("Download request was method %s; want GET", r.Method)
+		}
+		w.WriteHeader(http.StatusInternalServerError)
+	}))
+	defer httpServer.Close()
+	_, err := RetrieveTinkPublicKey(httpServer.URL)
+	if err == nil {
+		t.Errorf("Expected failure but call succeeded.")
+	}
+}
+
+func TestEncryptDocumentSuccess(t *testing.T) {
+	htmlStr := loadTestFileString("sample_encryption.html")
+	if htmlStr == "" {
+		t.Errorf("HTML file load failed.")
+	}
+	sr := strings.NewReader(googPublicKeyStr)
+	r := keyset.NewJSONReader(sr)
+	ks, err := r.Read()
+	if err != nil {
+		t.Errorf("Keyset load failed.")
+	}
+	pubKeys := map[string]tinkpb.Keyset{
+		"google.com": *ks,
+	}
+	encDoc, err := GenerateEncryptedDocument(htmlStr, "norcal.com:premium", pubKeys)
+	if err != nil {
+		t.Errorf("Error occured generating encrypted document.")
+	}
+	if !strings.Contains(encDoc, `<script type="application/octet-stream" ciphertext="">`) {
+		t.Errorf("Missing encrypted script.")
+	}
+	if !strings.Contains(encDoc, `<script type="application/json" cryptokeys="">`) {
+		t.Errorf("Missing cryptokeys script.")
+	}
+	if !strings.Contains(encDoc, "google.com") {
+		t.Errorf("Missing google.com key.")
+	}
+}
+
+func TestEncryptDocumentEmptyKeyset(t *testing.T) {
+	htmlStr := loadTestFileString("sample_encryption.html")
+	if htmlStr == "" {
+		t.Errorf("HTML file load failed.")
+	}
+	pubKeys := map[string]tinkpb.Keyset{
+		"google.com": tinkpb.Keyset{},
+	}
+	_, err := GenerateEncryptedDocument(htmlStr, "norcal.com:premium", pubKeys)
+	if err == nil {
+		t.Errorf("Error did not occur on empty Keyset.")
+	}
+}
```

Comment:
For thoroughness, add a test for malformed/unparseable HTML. And a test if there are no encrypted sections (meaning no key needs to be added to head).

Reply:
I couldn't find a good failure case for unparsable HTML because the Go parser adds in a lot of missing tags when it can't find them. https://godoc.org/golang.org/x/net/html#Parse

I added a test for the case where there aren't any valid encryptable sections and when the content is in Hebrew (i.e. non-ASCII).